libfreerdp/crypto/crypto.c

```diff
@@ -288,88 +295,452 @@ char* crypto_cert_subject_common_name(X509* xcert, int* length)
+GEN_EDIPARTY, GEN_RID, GEN_IPADD (the later can contain nul-bytes).
+
+
+mapper(name, data, index, count)
+
+The mapper is passed:
+ - the GENERAL_NAME selected,
+ - the data,
+ - the index of the general name in the subjectAltNames,
+ - the total number of names in the subjectAltNames.
+
+The last parameter let's the mapper allocate arrays to collect objects.
+Note: if names are filtered,  not all the indices from 0 to count-1 are
+passed to mapper,  only the indices selected.
+
+When the mapper returns 0, map_subject_alt_name stops the iteration immediately.
+
+*/
+
+#define GEN_ALL (-1)
+
+typedef int (*general_name_mapper_pr)(GENERAL_NAME* name, void* data, int index, int count);
+
+static void map_subject_alt_name(X509* x509, int general_name_type, general_name_mapper_pr mapper,
+                                 void* data)
 {
 	int i;
+	int num;
+	STACK_OF(GENERAL_NAME) *gens;
+	gens = X509_get_ext_d2i(x509, NID_subject_alt_name, NULL, NULL);
 
-	free(lengths);
+	if (!gens)
+	{
+		return;
+	}
+
+	num = sk_GENERAL_NAME_num(gens);
 
-	if (alt_name)
+	for (i = 0; (i < num); i++)
 	{
-		for (i=0; i<count; i++)
+		GENERAL_NAME* name = sk_GENERAL_NAME_value(gens, i);
+
+		if (name)
 		{
-			if (alt_name[i])
-				OPENSSL_free(alt_name[i]);
+			if ((general_name_type == GEN_ALL) || (general_name_type == name->type))
+			{
+				if (!mapper(name, data, i, num))
+				{
+					break;
+				}
+			}
 		}
+	}
+
+	sk_GENERAL_NAME_pop_free(gens, GENERAL_NAME_free);
+}
+
+
+/*
+extract_string  --  string extractor
+
+- the strings array is allocated lazily, when we first have to store a
+  string.
+
+- allocated contains the size of the strings array, or -1 if
+  allocation failed.
+
+- count contains the actual count of strings in the strings array.
+
+- maximum limits the number of strings we can store in the strings
+  array: beyond, the extractor returns 0 to short-cut the search.
+
+extract_string stores in the string list OPENSSL strings,
+that must be freed with OPENSSL_free.
+
+*/
+
+typedef struct string_list
+{
+	char**   strings;
+	int      allocated;
+	int      count;
+	int      maximum;
+} string_list;
+
+static void string_list_initialize(string_list* list)
+{
+	list->strings = 0;
+	list->allocated = 0;
+	list->count = 0;
+	list->maximum = INT_MAX;
+}
 
-		free(alt_name);
+static void string_list_allocate(string_list* list, int allocate_count)
+{
+	if (!list->strings && list->allocated == 0)
+	{
+		list->strings = calloc(allocate_count, sizeof(list->strings[0]));
+		list->allocated = list->strings ? allocate_count : -1;
+		list->count = 0;
 	}
 }
 
-char** crypto_cert_subject_alt_name(X509* xcert, int* count, int** lengths)
+static void string_list_free(string_list* list)
 {
-	int index;
-	int length = 0;
-	char** strings = NULL;
-	BYTE* string;
-	int num_subject_alt_names;
-	GENERAL_NAMES* subject_alt_names;
-	GENERAL_NAME* subject_alt_name;
+	// Note: we don't free the contents of the strings array: this
```

Comment:
Please no C++ comments

Reply:
ok.
By the way, there are 
find . -name \*.[hc] -exec grep -nH [^:]// {} +|grep -v '\/\*'|wc -l
1114 occurences of C++ comment in this code base!